Fogl 2655 - Plugins snapshot

amarendra-dianomic opened the pull request:
```
$ curl -X GET http://localhost:8081/foglamp/snapshot/plugins | jq
{
  "snapshots": []
}

$ curl -X POST http://localhost:8081/foglamp/snapshot/plugins | jq
{
  "message": "snapshot id=1554808178, file=snapshot-plugin-1554808178.tar.gz created successfully."
}

$ curl -X GET http://localhost:8081/foglamp/snapshot/plugins | jq
{
  "snapshots": [
    {
      "id": "1554808178",
      "name": "snapshot-plugin-1554808178.tar.gz"
    }
  ]
}

$ rm -rf C/plugins
$ ll C/plugins
ls: cannot access 'C/plugins': No such file or directory

$ curl -X PUT http://localhost:8081/foglamp/snapshot/plugins/1554808178 | jq
{
  "message": "snapshot snapshot-plugin-1554808178.tar.gz restored successfully."
}

$ ll C/plugins
total 28
drwxrwxr-x 7 asinha asinha 4096 Mar 25 11:52 ./
drwxrwxr-x 8 asinha asinha 4096 Apr  9 16:42 ../
drwxrwxr-x 3 asinha asinha 4096 Apr  8 15:25 common/
drwxrwxr-x 3 asinha asinha 4096 Mar 25 11:52 filter/
drwxrwxr-x 4 asinha asinha 4096 Mar 25 11:52 north/
drwxrwxr-x 6 asinha asinha 4096 Apr  3 17:13 storage/
drwxrwxr-x 2 asinha asinha 4096 Apr  3 17:13 utils/
$ 

$ curl -X DELETE http://localhost:8081/foglamp/snapshot/plugins/1554808178 | jq
{
  "message": "snapshot snapshot-plugin-1554808178.tar.gz deleted successfully."
}

$ curl -X GET http://localhost:8081/foglamp/snapshot/plugins | jq
{
  "snapshots": []
}
```

amarendra-dianomic commented:
@MarkRiddoch sure will test this PR against an installed environment like RPi before merging.

amarendra-dianomic commented:
@MarkRiddoch yes, code needed some fixes for location of C plugins based on FOGLAMP_ROOT variable. Now for dev environment, C plugins are being picked from $FOGLAMP_ROOT/C/plugins and for production environment, they are being picked from $FOGLAMP_ROOT/plugins. Python plugins in both the cases, reside in $FOGLAMP_ROOT/python/foglamp/plugins, so no issues. 

But there is a permission issue when we try to restore a plugin snapshot in production:

```
curl -X PUT http://localhost:8081/foglamp/plugins/snapshot/1554291387
500: Snapshot snapshot-plugin-1554291387.tar.gz could not be restored. Extraction error for snapshot /usr/local/foglamp/data/snapshots/plugins/snapshot-plugin-1554291387.tar.gz. [Errno 13] Permission denied: '/usr/local/foglamp/python/foglamp/plugins/north/ocs/ocs.py
```
I think, in production, it will probably be a manual restore of a snapshot plugin. Please examine.

MarkRiddoch commented:
> I think, in production, it will probably be a manual restore of a snapshot plugin. Please examine.

The use of this will be from the management system, it will allow us to undo changes to the installed plugins. So we will need to do this without manual intervention. I suggest we have a suid script to do it.

praveen-garg commented on python/foglamp/services/core/api/snapshot_plugin.py:
```diff
@@ -0,0 +1,116 @@
+
+
+__author__ = "Amarendra K Sinha"
+__copyright__ = "Copyright (c) 2017 OSIsoft, LLC"
```
fix the copyright please

praveen-garg commented on python/foglamp/services/core/api/snapshot_plugin.py:
```diff
@@ -0,0 +1,116 @@
+_help = """
+    -------------------------------------------------------------------------------
+    | GET POST        | /foglamp/plugins/snapshot                                 |
+    | PUT DELETE      | /foglamp/plugins/snapshot/{id}                            |
```
fix the routes please

praveen-garg commented on python/foglamp/services/core/api/snapshot_plugin.py:
```diff
@@ -0,0 +1,116 @@
+    snapshot_name = "snapshot-plugin-{}.tar.gz".format(snapshot_id)
+
+    if not os.path.isdir(_get_snapshot_dir()):
+        raise web.HTTPNotFound(reason="Snapshot directory does not exist")
```
better to return `reason="No snapshot found."` and put a (sys) log that directory does not exist

praveen-garg commented on python/foglamp/services/core/api/snapshot_plugin.py:
```diff
@@ -0,0 +1,116 @@
+    snapshot_name = "snapshot-plugin-{}.tar.gz".format(snapshot_id)
+
+    if not os.path.isdir(_get_snapshot_dir()):
+        raise web.HTTPNotFound(reason="Snapshot directory does not exist")
```
as above

praveen-garg commented on python/foglamp/services/core/snapshot.py:
```diff
@@ -0,0 +1,97 @@
+__version__ = "${VERSION}"
+
+_LOGGER = logger.setup(__name__)
+_NO_OF_FILES_TO_RETAIN = 3
```
why?  I can imagine some limit but not sure if that is 3? /cc @MarkRiddoch

MarkRiddoch replied:
This was not part of the original requirement, but is a good idea. The intended use is that there will probably only be a single snapshot, so 3 is as good a choice as any currently. Let's leave it at this, we should clearly indicate when we purge old ones in the log so that it is obvious if/when we get any issues form this.

praveen-garg commented on python/foglamp/services/core/snapshot.py:
```diff
@@ -0,0 +1,97 @@
+    async def build(self):
+        try:
+            snapshot_id = str(int(time.time()))
+            tar_file_name = self._out_file_path+"/"+"snapshot-plugin-{}.tar.gz".format(snapshot_id)
```
please don't mix + and format, use any. and if `+` use space around operator

praveen-garg commented on python/foglamp/services/core/snapshot.py:
```diff
@@ -0,0 +1,97 @@
+            pyz = tarfile.open(tar_file_name, "w:gz")
+            try:
+                pyz.add("{}/python/foglamp/plugins".format(_FOGLAMP_ROOT), recursive=True)
+                if _FOGLAMP_ROOT == '/usr/local/foglamp':
```
Please add a comment that it is for C plugins path for `make` vs `make install`

praveen-garg commented on python/foglamp/services/core/snapshot.py:
```diff
@@ -0,0 +1,97 @@
+            raise RuntimeError(str(ex))
+
+        self.check_and_delete_temp_files(self._interim_file_path)
+        _LOGGER.info("Snapshot bundle %s successfully created.", tar_file_name)
```
remove `bundle` ?

praveen-garg commented on python/foglamp/services/core/routes.py:
```diff
@@ -175,6 +176,12 @@ def setup(app):
+    app.router.add_route('GET', '/foglamp/plugins/snapshot', snapshot_plugin.get_snapshot)
+    app.router.add_route('POST', '/foglamp/plugins/snapshot', snapshot_plugin.post_snapshot)
+    app.router.add_route('PUT', '/foglamp/plugins/snapshot/{id}', snapshot_plugin.put_snapshot)
+    app.router.add_route('DELETE', '/foglamp/plugins/snapshot/{id}', snapshot_plugin.delete_snapshot)
```
fix the routes please

ashish-jabble commented on python/foglamp/services/core/api/snapshot/plugins.py:
```diff
@@ -0,0 +1,118 @@
+    if _FOGLAMP_DATA:
+        snapshot_dir = os.path.expanduser(_FOGLAMP_DATA + '/snapshots/plugins')
+    else:
+        snapshot_dir = os.path.expanduser(_FOGLAMP_ROOT + '/data/snapshots/plugins')
```
`data/plugins/snapshot` and please add `data/plugins` in .gitignore file

amarendra-dianomic replied:
done

amarendra-dianomic commented:
@MarkRiddoch now extract plugins snapshot .gz file is working fine with "sudo make install" also. For this, I changed the extract method to call the external command "tar" instead of using Python tarfile module because of permission issues. 

However, we will need to add appropriate instructions in  /etc/sudoers file to suppress sudo password prompt during the rest api call. I have added a command in scripts/extras/sudoers.foglamp. But as this file will be used only when installing foglamp via a .deb package, so in order to check it locally with "sudo make install", it will be required to manually copy content of this file to /etc/sudoers via `sudo visudo` command.

@praveen-garg all feedback changes done. Pl check.

ashish-jabble commented on python/foglamp/services/core/api/snapshot/plugins.py:
```diff
@@ -0,0 +1,118 @@
+
+
+__author__ = "Amarendra K Sinha"
+__copyright__ = "Copyright (c) 2019 Diaonmic Systems"
```
Dianomic

ashish-jabble commented on scripts/extras/foglamp.sudoers:
```diff
@@ -1,2 +1,3 @@
-%sudo ALL=(ALL) NOPASSWD: /usr/bin/apt-get -y update, /usr/bin/apt-get -y install foglamp
-
+%sudo ALL=(ALL) NOPASSWD: /usr/bin/apt-get -y update
+%sudo ALL=(ALL) NOPASSWD: /usr/bin/apt-get -y install foglamp
```
Use `apt` instead of `apt-get` as apt is more structured and provides you with necessary options needed to manage packages.

amarendra-dianomic replied:
Let us take up this improvement separately as it really belongs to foglamp-pk repo and it will need thorough testing with deb packages.

MarkRiddoch replied:
We could drop tis file from the PR altogether now (and update the description of the PR), since we are to unsung sudo any more for this.

ashish-jabble commented on python/foglamp/services/core/snapshot.py:
```diff
@@ -0,0 +1,116 @@
+                if _FOGLAMP_ROOT == '/usr/local/foglamp':
+                    pyz.add("{}/plugins".format(_FOGLAMP_ROOT), recursive=True, filter=reset)
+                else:
+                    pyz.add("{}/C/plugins".format(_FOGLAMP_ROOT), recursive=True)
```
`C-plugins` location is wrong when make, it is under `FOGLAMP_ROOT/cmake_build/C/plugins/`.

And to avoid different build process check please use `FOGLAMP_ROOT/plugins` symlink directory; as we have already discovered C-plugins from the same location.

amarendra-dianomic replied:
C plugins snapshot actually means creating snapshot of the C source and NOT the final build artifacts. Hence the reason for selecting FOGLAMP_ROOT/C/plugins as C plugins location. Whenever we will restore a plugin snapshot, we will be required to run make/make install again.

ashish-jabble replied:
As such no source file exists for external plugins ONLY compiled. So for `make` process we need to add below combination
```
FOGLAMP_ROOT/cmake_build/C/plugins/ -- Compiled library
FOGLAMP_ROOT/plugins --- SYMLINK
```

amarendra-dianomic commented on python/foglamp/services/core/api/snapshot/plugins.py:
```diff
@@ -0,0 +1,118 @@
+    snapshot_dir = _get_snapshot_dir()
+    valid_extension = '.tar.gz'
+    found_files = []
+    find_id = lambda x: x.split("snapshot-plugin-")[1].split(".tar.gz")[0]
```
why?

MarkRiddoch commented on python/foglamp/services/core/snapshot.py:
```diff
@@ -0,0 +1,118 @@
+
+    def extract_files(self, pyz):
+        # Since we are storing full path of the files, we need to specify "/" as the path to restore
+        if _FOGLAMP_ROOT == '/usr/local/foglamp':
```
I probably was not very clear, when I said a suid script what I meant was a shell script that was set user id on execution, i.e. chmod 4755 .. That way we can make the script very limited in what is does, i.e. only restore a snapshot from the fixed snapshot directory.

amarendra-dianomic replied:
Thanks @MarkRiddoch a new script scripts/restore_plugins_snapshot has been added with setuid and setgid bits being set during "sudo make install". This has been successfully tested. Hence new sudoers entry for /bin/tar has been removed.

MarkRiddoch commented on scripts/extras/foglamp.sudoers:
```diff
@@ -1,2 +1,3 @@
-
+%sudo ALL=(ALL) NOPASSWD: /usr/bin/apt-get -y update
+%sudo ALL=(ALL) NOPASSWD: /usr/bin/apt-get -y install foglamp
+%sudo ALL=(ALL) NOPASSWD: /bin/tar -C / -xzf *
```
If we use the set uid script we should not need this and it will be more secure.

amarendra-dianomic replied:
Correct. This entry now stands removed. Please see comment above.

amarendra-dianomic commented:
@MarkRiddoch @praveen-garg @ashish-jabble all feedback changes done and tested ok. Please check.

MarkRiddoch commented on scripts/restore_plugins_snapshot:
```diff
@@ -0,0 +1,34 @@
+  FILE_NAME="$1"
+fi
+
+sudo /bin/tar -C / -xzf "$FILE_NAME"
```
If this is a setuid on execution script we do not need the sudo in this line.

amarendra-dianomic replied:
Regret to inform that this setuid thing is not working. The earlier "success" was due to the fact that the script restore_snapshot_plugins contained a sudo /bin/tar command and before executing this script, I had probably run a sudo command due to which the sudo password remained active for some time. I deeply regret the slip.

Secondly, the setuid for a script is considered a security risk and there are plenty of reference material available on the net advising against it, A few such reference urls are:
https://unix.stackexchange.com/questions/364/allow-setuid-on-shell-scripts
http://www.faqs.org/faqs/unix-faq/faq/part4/section-7.html

Thereafter, I decided to test the setuid bit on a test script as below. I created a folder test, owned by root, and created a shell script, also owned by root but having its setuid bit set. When executed by a normal user, asinha, it did not work and threw permission error. 
```
~ $ whoami
asinha
~ $ 
~ $ ll -d test
drwxr-xr-x 2 root root 4096 Apr 11 21:07 test/
~ $ 
~ $ cd test
test $ 
test $ ll
total 12
drwxr-xr-x  2 root   root   4096 Apr 11 21:07 ./
drwxr-xr-x 32 asinha asinha 4096 Apr 11 21:06 ../
-rwSr-Sr-x  1 root   root     34 Apr 11 21:00 scrpt*
test $ 
test $ cat scrpt 
#!/bin/bash
mkdir test2
cd test2

test $ 
test $ whoami
asinha
test $ 
test $ ./scrpt 
mkdir: cannot create directory ‘test2’: Permission denied
./scrpt: line 3: cd: test2: No such file or directory
test $ 
test $
```
Therefore, it seems that we have no other option but to revert back to entries in foglamp.sudoers file. Please advise.

AmandeepArora replied:
We can probably write a small C helper program to do tar archive extract in '/' and set setuid bit on it.

amarendra-dianomic replied:
@AmandeepArora to check this, can we create a quick C executable to replicate the scrpt script file above?

amarendra-dianomic replied:
@AmandeepArora @MarkRiddoch a C++ executable, test, provided by @AmandeepArora seems to be working fine and thus serves our purpose as below:
```
FogLAMP (FOGL-2655 %) $ ll -d test
drwxr-xr-x 2 root root 4096 Apr 15 12:03 test/
FogLAMP (FOGL-2655 %) $ 
FogLAMP (FOGL-2655 %) $ 
FogLAMP (FOGL-2655 %) $ cd test
test (FOGL-2655 %) $ 
test (FOGL-2655 %) $ 
test (FOGL-2655 %) $ ll
total 20
drwxr-xr-x  2 root   root   4096 Apr 15 12:03 ./
drwxrwxr-x 23 asinha asinha 4096 Apr 15 11:55 ../
-rwsr-sr-x  1 root   root   8824 Apr 15 11:55 test*
test (FOGL-2655 %) $ 
test (FOGL-2655 %) $ 
test (FOGL-2655 %) $ whoami
asinha
test (FOGL-2655 %) $ 
test (FOGL-2655 %) $ 
test (FOGL-2655 %) $ ./test
test (FOGL-2655 %) $ 
test (FOGL-2655 %) $ 
test (FOGL-2655 %) $ ll
total 24
drwxr-xr-x  3 root   root   4096 Apr 15 12:04 ./
drwxrwxr-x 23 asinha asinha 4096 Apr 15 11:55 ../
-rwsr-sr-x  1 root   root   8824 Apr 15 11:55 test*
drwxrwx---  2 root   root   4096 Apr 15 12:04 test2/
test (FOGL-2655 %) $ 
test (FOGL-2655 %) $ 
test (FOGL-2655 %) $ whoami
asinha
test (FOGL-2655 %) $ 
```

Hence it is proposed to go ahead with the C helper program to extract a .tar.gz file as proposed by @AmandeepArora above. This C++ helper program will receive .tar.gz filename as a command line argument.

Thanks @AmandeepArora for this initiative.

AmandeepArora replied:
I provided this C binary for tar extraction. There is still a major security risk involved. The program would extract the compressed archive by setting "-C /", so in a way, it could potentially be used to even overwrite binaries in /bin, /usr/bin/, /sbin, etc. May be there should be further verification of tar archive contents in this C binary program before extraction.

MarkRiddoch replied:
Good, let's go with a small utility program to do it. Seems better than having to use sudo to me. Seems much more secure.

ashish-jabble commented on Makefile:
```diff
@@ -184,7 +184,8 @@ install : $(INSTALL_DIR) \
 	extras_install \
-	data_install
+	data_install \
+	setuid_extract_plugin_snapshot
```
This target should call inside

```
extras_install : $(EXTRAS_INSTALL_DIR) install_python_fogbench install_extras_scripts setuid_extract_plugin_snapshot
```

amarendra-dianomic replied:
@ashish-jabble can you please elaborate upon the shortcomings/errors in the current code and how the proposed code solves/fixes that?

ashish-jabble replied:
As extract_plugin_snapshot is a utility and the part of `extras/`. In _Makefile_ we made as a practice all extras utilities should go under `extras_install`. So that it will clear to all what are the extras we made for foglamp.

ashish-jabble commented on Makefile:
```diff
@@ -442,11 +443,16 @@ endif
 
+# set setuid bit of extract_plugin_snapshot
+setuid_extract_plugin_snapshot:
+	chmod u+s $(INSTALL_DIR)/extras/C/extract_plugin_snapshot
```
`chmod u+s $(EXTRAS_INSTALL_DIR)/C/extract_plugin_snapshot`

amarendra-dianomic replied:
@ashish-jabble can you please elaborate upon the shortcomings/errors in the current code and how the proposed code solves/fixes that?

ashish-jabble replied:
Reuse `$(EXTRAS_INSTALL_DIR)` it. In future if we want to change the directory structure so we need to change at one place NOT on basis of find occurrence.

praveen-garg commented on scripts/restore_plugins_snapshot:
```diff
@@ -0,0 +1,41 @@
+#!/bin/bash
```
do we still need this script?

praveen-garg commented on python/foglamp/services/core/snapshot.py:
```diff
@@ -0,0 +1,121 @@
+                sorted_list = sorted(list_files, key=lambda k: k['id'], reverse=True)
+                if len(sorted_list) >= _NO_OF_FILES_TO_RETAIN:
+                    for f in sorted_list[_NO_OF_FILES_TO_RETAIN-1:]:
+                        os.remove(os.path.join(snapshot_plugin_dir, f['name']))
```
Please add log that we are purging the oldest having name x

praveen-garg commented on python/foglamp/services/core/api/snapshot/plugins.py:
```diff
@@ -0,0 +1,166 @@
+                           "name": x}, valid_files))
+            sorted_list = sorted(list_files, key=lambda k: k['id'], reverse=True)
+    except:
+        pass
```
Wait, when we got this added?! Why a pass
